analyzer/src/types/ctx.rs

```diff
@@ -0,0 +1,69 @@
+use crate::relations::{Relations, SourceObjectId, Symbol};
+use crate::types::hir::TypeId;
+use crate::types::{BOOL, FLOAT, INT, NOTHING, STRING};
+use std::collections::HashMap;
+
+pub struct TypeContext {
+    pub(crate) source: SourceObjectId,
+    names: HashMap<String, TypeId>,
+    locals: HashMap<SourceObjectId, Vec<TypeId>>,
+}
+
+impl TypeContext {
+    pub(crate) fn lang() -> Self {
+        Self {
+            source: SourceObjectId(0),
+            names: HashMap::from([
+                ("Nothing".to_owned(), NOTHING),
+                ("Bool".to_owned(), BOOL),
+                ("Int".to_owned(), INT),
+                ("Float".to_owned(), FLOAT),
+                ("String".to_owned(), STRING),
+            ]),
+            locals: HashMap::new(),
+        }
+    }
```

Comment:
i think you should rather have a `TypeContext::new` function, and fill-in the lang types elsewhere, this struct is a struct for all the type analysis process and is meant to mutate a lot, thus not really bound to an initial state. also `source: SourceObjectId` is weird, i suppose that it's just here as a default value as the context seems to be _prepared_ before any real use

```suggestion
    pub(crate) fn new(initial_source: SourceObjectId) -> Self {
        let slf = Self {
            source: initial_source,
            names: HashMap::empty(),
            locals: HashMap::new(),
        }
        slf.prepare(initial_source);
        slf
    }
```